kernel/src/hil/framebuffer.rs

```diff
@@ -0,0 +1,57 @@
+//! Hil for FrameBuffer
+use crate::returncode::ReturnCode;
+use enum_primitive::cast::FromPrimitive;
+use enum_primitive::enum_from_primitive;
+
+enum_from_primitive! {
+    #[derive(Copy, Clone, PartialEq)]
+    pub enum ScreenRotation {
+        Normal = 0,
+        Rotated90 = 1,
+        Rotated180 = 2,
+        Rotated270 = 3,
+    }
+}
+
+impl From<ScreenRotation> for usize {
+    fn from(rotation: ScreenRotation) -> usize {
+        match rotation {
+            ScreenRotation::Normal => 0,
+            ScreenRotation::Rotated90 => 1,
+            ScreenRotation::Rotated180 => 2,
+            ScreenRotation::Rotated270 => 3,
+        }
+    }
+}
+
+pub trait Screen {
+    fn set_resolution(&self, width: usize, height: usize) -> ReturnCode;
+    fn set_color_depth(&self, depth: usize) -> ReturnCode;
+    fn set_rotation(&self, rotation: ScreenRotation) -> ReturnCode;
+
+    fn get_resolution(&self) -> (usize, usize);
+    fn get_color_depth(&self) -> usize;
+    fn get_rotation(&self) -> ScreenRotation;
+
+    fn get_resolution_modes(&self) -> usize;
+    fn get_resolution_size(&self, index: usize) -> (usize, usize);
+
+    fn get_color_depth_modes(&self) -> usize;
+    fn get_color_depth_bits(&self, index: usize) -> usize;
+
+    fn write(&self, x: usize, y: usize, width: usize, height: usize) -> ReturnCode;
```

Comment:
How does this interface work?

Reply:
``get_color_depth_modes`` - returns the count of the color depths supported
``get_color_depth_bits`` - can be called with index 0 .. count-1 and will return the value of each color depth mode
``write`` - sends a write command to write data in the selected memory window, it seems that several screens can do that. The screen will then call ``fill_next_buffer_for_write`` from the ``ScreenClient`` for subsequent writes. This function will fill the buffer and return the number of bytes written. If it returns 0, the write stops and the screen issues ``command_complete``. This was there is no need for triple buffering (an app buffer, a frame buffer buffer and a screen buffer), data is transfered from the app directly. This also allows writing a repeated pattern (for now only one color) without the app having to fill a buffer with the same pattern. It also allow the screen to have internal an arbitrary size buffer.

Reply:
It would be great to put this documentation right in the HIL file.

Reply:
I added it to the HIL file.

Reply:
I changed the way the buffer is used, now the buffer is passed down from the frame buffer driver to the screen driver.